File: src/com/github/luomingxu/idea/util/ExcludeFileSearch.java
package com.github.luomingxu.idea.util;

import com.intellij.openapi.vfs.VirtualFile;

import java.io.File;

public class ExcludeFileSearch {

    static class MsgException extends Exception {
        public String msg;

        public MsgException(String msg) {
            this.msg = msg;
        }
    }

    private static void searchFileByName(File file, String fileName) throws Exception {
        if (file.isDirectory()) {
            for (File item : file.listFiles()) {
                searchFileByName(item, fileName);
            }
        } else {
            if (file.getName().equals(fileName)) {
                throw new MsgException(file.getCanonicalPath());
            }
        }
    }

    public static String getFileByName(String fileName, VirtualFile[] excludeDirs) {
        for (VirtualFile item : excludeDirs) {
            try {
                searchFileByName(new File(item.getCanonicalPath()), fileName);
            } catch (Exception e) {
                if (e instanceof MsgException) {
                    return ((MsgException) e).msg;
                } else {
                    e.printStackTrace();
                }
            }
        }

        return null;
    }
}

File: src/main/java/com/github/luomingxu/idea/util/ExcludeFileSearch.kt
package com.github.luomingxu.idea.util

import com.intellij.openapi.vfs.VirtualFile
import java.io.File

object ExcludeFileSearch {
    @Throws(MsgException::class)
    private fun searchFileByName(file: File, fileName: String) {
        if (file.isDirectory) {
            file.listFiles()?.let {
                for (item in it) {
                    searchFileByName(item, fileName)
                }
            }
        } else if (file.name == fileName) {
            throw MsgException(file.canonicalPath)
        }
    }

    fun getFileByName(fileName: String, excludeDirs: Array<VirtualFile>): String? {
        for (item in excludeDirs) {
            try {
                item.canonicalPath?.let {
                    searchFileByName(File(it), fileName)
                }
            } catch (e: Exception) {
                if (e is MsgException) {
                    return e.msg
                } else {
                    e.printStackTrace()
                }
            }
        }
        return null
    }

    internal class MsgException(var msg: String) : Exception()
}
